### PR TITLE
[store] Add response hints to label APIs

### PR DESCRIPTION
**Pull request by gouthamve**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

In Cortex, we shard blocks b/w store-gateways, hence we need to send the list of blockids to query and the list of ids that have been queried in the response to make sure all the relevant blocks are queried. This has already been done for the `Series` API.

I have not yet added request hints yet. I will add them in a subsequent PR.

## Verification

tbh, I have not tested it yet. I am only relying on the unit tests. Any tips on how I can test it?

### CODE REVIEW

File: pkg/store/bucket_test.go [hunk 1]
```diff
@@ -1897,3 +1897,190 @@ func createBlockWithOneSeriesWithStep(t testutil.TB, dir string, lbls labels.Lab
...
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			namesResp, err := store.LabelNames(context.Background(), tc.labelNamesReq)
+			testutil.Ok(t, err)
+			testutil.Equals(t, tc.expectedNames, namesResp.Names)
```
Comment (gouthamve) [inline pkg/store/bucket_test.go:2055]: This fails with:
```
        bucket_test.go:2055: bucket_test.go:2055:
            
                exp: []string{"ext1", "foo", "i"}
            
                got: []string{"foo", "i"}
            
            Diff:
            --- Expected
            +++ Actual
            @@ -1,3 +1,2 @@
            -([]string) (len=3) {
            - (string) (len=4) "ext1",
            +([]string) (len=2) {
              (string) (len=3) "foo",
```


As in, the `extLabels` we passed are not being returned. From what I know, they should be, but I am not even sure what they are. /cc @pracucci @pstibrany

**Review by pracucci — APPROVED** [review]
Thanks @gouthamve! LGTM. Could you take a look at the comments, fix tests and rebase `master`, please?

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -1070,13 +1072,27 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
...
 		indexr := b.indexReader(gctx)
 		g.Go(func() error {
 			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")
 
 			// Do it via index reader to have pending reader registered correctly.
 			res := indexr.block.indexHeaderReader.LabelNames()
+
+			// Add  a set for the external labels as well.
+			// We're not adding them directly to res because there could be duplicates.
+			extRes := make([]string, 0, len(b.meta.Thanos.Labels))
```
Comment (pracucci) [inline pkg/store/bucket.go:1087]: You can't safely access `b` inside the goroutine.

File: pkg/store/bucket.go [hunk 3]
```diff
@@ -1119,6 +1150,11 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 				return errors.Wrap(err, "index header label values")
 			}
 
+			// Add the external label value as well.
+			if extLabelValue, ok := b.meta.Thanos.Labels[req.Label]; ok {
```
Comment (pracucci) [inline pkg/store/bucket.go:1154]: You can't safely access `b` inside the goroutine.

**Comment by gouthamve** [comment]
Fixes https://github.com/thanos-io/thanos/issues/2984

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

File: pkg/store/bucket.go [hunk 4]
```diff
@@ -1102,13 +1119,25 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
...
 		return nil, status.Error(codes.Internal, err.Error())
 	}
+
+	var anyHints *types.Any
+	var err error
+
+	if anyHints, err = types.MarshalAny(resHints); err != nil {
+		err = status.Error(codes.Unknown, errors.Wrap(err, "marshal label names response hints").Error())
+		return nil, err
+	}
```
Comment (bwplotka) [inline pkg/store/bucket.go:1129]: Why declaring `anyHints` and err in separate lines? Why not:

```suggestion
	anyHints, err := types.MarshalAny(resHints); 
	if err != nil {
		return, status.Error(codes.Unknown, errors.Wrap(err, "marshal label names response hints").Error())
	}
```

File: pkg/store/bucket.go [hunk 5]
```diff
@@ -1143,8 +1182,18 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 	if err := g.Wait(); err != nil {
 		return nil, status.Error(codes.Aborted, err.Error())
 	}
+
+	var anyHints *types.Any
+	var err error
+
+	if anyHints, err = types.MarshalAny(resHints); err != nil {
```
Comment (bwplotka) [inline pkg/store/bucket.go:1189]: ditto

**Review by bwplotka — APPROVED** [review]
Nice, work! Minor style nit only, otherwise LGTM :+1:

**Comment by gouthamve** [comment]
@bwplotka Nice catch, not sure what I was thinking there. Fixed it!

**Review by bwplotka — APPROVED** [review]
LGTM, we can ignore Markdown check link, it's flaky, otherwise, we can merge on green

**Pull request merged**